### PR TITLE
Island: Remove trailing slashes before registering a URL

### LINKED ISSUE #2191: Bug in `AggregatingPropagationCredentialsRepository` causes agent to crash
## Describe the bug
This bug is twofold:
1. The `/api/propagation-credentials` endpoint is [defined with a trailing slash](https://github.com/guardicore/monkey/blob/956621dcf3a912c2463b748a596d30313e87421a/monkey/monkey_island/cc/resources/propagation_credentials.py#L14). This results in a 404 response when the agent attempts to get credentials from the Island, even though `strict_slashes=False` (https://stackoverflow.com/questions/33241050/trailing-slash-triggers-404-in-flask-path-rule)
2. The SSH exploiter fails with the following traceback:
```
 2022-08-11 23:32:03,239 [15:ExploiterThread-02:ERROR] HostExploiter.exploit_host.79: Exception in exploit_host
 Traceback (most recent call last):
   File \"infection_monkey/exploit/HostExploiter.py\", line 74, in exploit_host
     return self._exploit_host()
   File \"infection_monkey/exploit/sshexec.py\", line 165, in _exploit_host
     ssh = self.exploit_with_ssh_keys(port)
   File \"infection_monkey/exploit/sshexec.py\", line 49, in exploit_with_ssh_keys
     identities=self.options[\"credentials\"][\"exploit_user_list\"],
 TypeError: 'NoneType' object is not subscriptable
```
This is likely the result of a bug in the way that `AggregatingPropagationCredentialsRepository` is handling the 404 error from the credentials endpoint.

## To Reproduce
1. Run the `depth_2_a` test
1. See it fail
1. Inspect the agent logs (see screenshot below)

## Recommendation
1. Diagnose the `TypeError`. It's likely an issue with `AggregatingPropagationCredentialsRepository`.
2. Write a unit test to reproduce it.
4. Fix it.
5. Fix the `/api/propagation-credentials` endpoint so that the agent can successfully retrieve credentials.

## Screenshots
image
image

### PR DESCRIPTION
**Pull request by VakarisZ**
Strict slashes seems to not handle a case when URL is defined with a trailing slash, but request is sent without one. Removing trailing slashes before registering a URL will solve the burden of remembering to register URLS without slashes

# What does this PR do?

Fixes the 404 in  #2191

Add any further explanations here.

## PR Checklist
* [ ] Is the TravisCI build passing?
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?
* [ ] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by running the Agent

### CODE REVIEW

**Review by mssalvatore — APPROVED** [review]
I'd like to know why `strict_slashes` isn't behaving the way it's supposed to, but if that'll take too much time this is ok for now.

**Comment by mssalvatore** [comment]
Ignore the flake8 issue, but unit tests fail.

File: monkey/monkey_island/cc/app.py [hunk 1]
```diff
@@ -129,6 +129,13 @@ def add_resource(self, resource: Type[AbstractResource]):
 
         self._reserve_urls(resource.urls)
 
+        # enforce our rule that URLs should not contain a trailing slash
+        for url in resource.urls:
+            if url.endswith("/"):
+                raise ValueError(
+                    f"Resource {resource.__name__} has an invalid URL. A URL "
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/app.py:136]:
```suggestion
                    f"Resource {resource.__name__} has an invalid URL: A URL "
```

File: monkey/tests/unit_tests/monkey_island/cc/resources/test_propagation_credentials.py [hunk 2]
```diff
@@ -21,9 +21,9 @@
     _stolen_collection,
 )
 
-ALL_CREDENTIALS_URL = PropagationCredentials.urls[0]
-CONFIGURED_CREDENTIALS_URL = urljoin(ALL_CREDENTIALS_URL, _configured_collection)
-STOLEN_CREDENTIALS_URL = urljoin(ALL_CREDENTIALS_URL, _stolen_collection)
+ALL_CREDENTIALS_URL = PropagationCredentials.urls[0].rstrip("/")
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/monkey_island/cc/resources/test_propagation_credentials.py:24]: It shouldn't have a trailing slash, right?

```suggestion
ALL_CREDENTIALS_URL = PropagationCredentials.urls[0]
```

**Review by mssalvatore — CHANGES_REQUESTED** [review]
Let's move the changes to `AggregatingPropagationCredentialsRepository` to a different PR.

File: monkey/tests/unit_tests/monkey_island/cc/resources/test_propagation_credentials.py (hunk 2 above)
Comment (cakekoa) [reply monkey/tests/unit_tests/monkey_island/cc/resources/test_propagation_credentials.py:24]: Not anymore!

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**